pkg/graveler/committed/merge_iterator.go

```diff
@@ -68,75 +81,135 @@ func (d *compareIterator) handleConflict() bool {
 		d.err = graveler.ErrConflictFound
 		return false
 	}
-	d.val = d.diffIt.Value().Copy()
+	val, _ := d.diffIt.Value()
+	d.val = val.Copy()
 	d.val.Type = graveler.DiffTypeConflict
 	return true
 }
 
-func (d *compareIterator) Next() bool {
-	for {
-		// get next value or return with context/iterator error (if any)
+func (d *compareIterator) Step() bool {
+	hasNext := true
+	for hasNext {
 		select {
 		case <-d.ctx.Done():
 			d.err = d.ctx.Err()
 			return false
 		default:
-			if !d.diffIt.Next() {
-				d.err = d.diffIt.Err()
-				return false
-			}
-		}
-		val := d.diffIt.Value()
-		key := val.Key
-		typ := val.Type
-		baseVal, err := d.valueFromBase(key)
-		if err != nil {
-			d.err = err
-			return false
 		}
-		switch typ {
-		case graveler.DiffTypeAdded:
-			// exists on source, but not on dest
-			if baseVal == nil {
-				// added only on source
-				d.val = d.diffIt.Value().Copy()
-				return true
-			}
-			if !bytes.Equal(baseVal.Identity, val.Value.Identity) {
-				// removed on dest, but changed on source
-				return d.handleConflict()
-			}
-			continue
-		case graveler.DiffTypeChanged:
-			if baseVal == nil {
-				// added on dest and source, with different identities
-				return d.handleConflict()
+		val, rngDiff := d.diffIt.Value()
+		if rngDiff != nil {
```

Comment:
Not sure this (allowing a call to `.Value()` on an iterator that has ended) is part of the Go iterator contract; paging @nopcoder...

Reply:
At this point the iterator shouldn't end ( If it could, there is a bug here).
`Step` is called after the  `Next` or `NextRange` has returned true and runs in a loop that updates hasNext.
Obviously it's not clear... changing it.